### PR TITLE
Jetpack Licensing: Streamlined logic for license issuing and payment method requirements

### PR DESCRIPTION
**Pull request by heydemoura**
#### Proposed Changes

* Context: 1201801459945917-as-1202552164778579/f
* Implemented a long term solution for the license issuing process with the issue license step coming before adding a payment method;
* The license issuing process is now more streamlined and does not rely only on a single component.

**Regarding `useLicenseIssuing(selectedSite?, product?)` Hook**
* Extracted license issuing mutation logic from issue license form to a hook, which can be called when successfully adding a payment method;
* Renamed `hooks.ts` to `hooks.tsx`, it might look like a whole file of additions, but it is just a new hook;

**`IssueLicenseForm`**
* No longer issuing a license from the license issuing form if there's no valid payment method on file. Instead, the license is issued after successfully adding a payment method (It will issue the license and not move to the payment method addition form if there's already a credit card on file);
* Moved auxiliary functions `selectProductOptions` and `alphabeticallySortedProductOptions` to `partner-portal/utils` module;
* Cleanup of `IssueLicenseForm` component, so that license issuing logic is handled on `useLicenseIssuing` hook, leaving mostly UI rendering and events implementation.
* There is a new event being tracked, labeled as `calypso_partner_portal_issue_selection_submit` for tracking when a user selects a product to be licensed and moves on to issuing the license or being required to input a payment method.

**`PaymentMethodAdd`**
* When adding a payment method, the _Go Back_ button can go back to product selection screen if there's a product selection;
* When adding a payment method with a `product` argument in the URL, it will then issue a license for that product after successfully adding the payment method;

#### Testing Instructions

* Expected flows described here: pdpAdu-v0-p2

#### Screenshots
**Success notice when issuing a license**
image

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

### CODE REVIEW

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+						name: getProductTitle( selectedProduct.name ),
+						key: licenseKey,
+					},
+				} )
+			);
+		}
+		return page.redirect( '/dashboard' );
+	};
+
+	const requirePaymentMethod = useCallback( () => {
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: Exported callback to handle redirection to `/payment-methods/add` logic.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+			);
+
+			if ( sites > 0 ) {
+				nextStep = addQueryArgs(
+					{ key: license.license_key },
+					partnerPortalBasePath( '/assign-license' )
+				);
+			}
+
+			page.redirect( nextStep );
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx:213]: Removed the case that the user is redirected to `/payment-methods/add` as the following step, since this is not a real case anymore. Using `requirePaymentMethod()` on the component level will lead the user to `/payment-methods/add` before issuing a license.

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 3]
```diff
@@ -42,141 +29,17 @@ export default function IssueLicenseForm( {
...
-			// In order to make this a smoother experience for the user, we retry a couple of times silently if the
-			// error is missing_valid_payment_method but our local state shows that the user has a payment method.
-			if ( ! paymentMethodRequired && error?.code === 'missing_valid_payment_method' ) {
-				return errorCount < 5;
-			}
-
-			return false;
-		},
-	} );
+	const [ issueLicense, isSubmitting, requirePaymentMethod ] = useLicenseIssuing(
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: The above was moved to the `useLicenseIssuing` hook

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 4]
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
 
 	const onIssueLicense = useCallback( () => {
 		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
-		issueLicense.mutate( { product } );
-	}, [ dispatch, product, issueLicense.mutate ] );
 
-	const selectedSiteDomian = selectedSite?.domain;
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx:211]: Fixed a typo

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 5]
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
...
-	useEffect( () => {
-		if ( defaultProduct !== '' ) {
-			page.redirect(
-				removeQueryArgs( window.location.pathname + window.location.search, 'product' )
-			);
-			issueLicense.mutate( { product: defaultProduct } );
+		if ( hasPaymentMethod ) {
+			issueLicense.mutate( { product } );
+		} else {
+			requirePaymentMethod();
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: If there's a payment method, it will issue the license already.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 6]
```diff
@@ -59,11 +64,18 @@ function PaymentMethodAdd(): ReactElement {
...
+
+	const product = useMemo(
+		() => ( getQueryArg( window.location.href, 'product' ) || '' ).toString(),
 		[ window.location.href, getQueryArg ]
 	);
 
-	useReturnUrl( hasPaymentMethod );
+	const [ issueLicense, isSubmitting ] = useLicenseIssuing( null, product );
+
+	useReturnUrl( !! returnQueryArg && hasPaymentMethod );
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Only redirects in case there's a `return` query argument. So it will not redirect on the regular case, that there's only a `product` query argument or no argument at all. The `return` query argument will only be present when the user tries to assign a license without a payment method on file.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 7]
```diff
@@ -96,7 +108,7 @@ function PaymentMethodAdd(): ReactElement {
 	);
 
 	const successCallback = useCallback( () => {
-		if ( returnQueryArg ) {
+		if ( returnQueryArg || product ) {
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:116]: Refetch the cards to update the `has_valid_payment_method` flag when a product is selected.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 8]
```diff
@@ -106,7 +118,13 @@ function PaymentMethodAdd(): ReactElement {
...
 			page( '/partner-portal/payment-methods/' );
 		}
-	}, [ reduxDispatch, page, window, getQueryArg ] );
+	}, [ page, product, issueLicense ] );
+
+	useEffect( () => {
+		if ( product && hasPaymentMethod ) {
+			issueLicense.mutate( { product } );
+		}
+	}, [ hasPaymentMethod, product ] );
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Issue the license once there's a product and a payment method on file.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 9]
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
...
 								<Button
 									className="payment-method-add__back-button"
-									href="/partner-portal/payment-methods/"
-									disabled={ isStripeLoading }
+									href={
+										product
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
+											: '/partner-portal/payment-methods/'
+									}
+									disabled={ isStripeLoading || issueLicense.isLoading || isSubmitting }
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: "Go Back" button will lead to the products list with the selected product highlighted, if there's a product selected, keeping the user on the license issuing flow.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx [hunk 10]
```diff
@@ -3,24 +3,30 @@
...
 
 		const { container, getByText } = render(
-			<Provider store={ store }>
-				<PaymentMethodAdd />
-			</Provider>
+			<QueryClientProvider client={ queryClient }>
+				<Provider store={ store }>
+					<PaymentMethodAdd />
+				</Provider>
+			</QueryClientProvider>
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/test/payment-method-add.jsx:29]: Now that licenses can be issued from the this component, these mocks were needed for testing.

File: client/jetpack-cloud/sections/partner-portal/utils.ts [hunk 11]
```diff
@@ -176,3 +182,13 @@ export function formatApiPartner( partner: APIPartner ): Partner {
...
+
+export function selectProductOptions( families: APIProductFamily[] ): APIProductFamilyProduct[] {
+	return families.flatMap( ( family ) => family.products );
+}
+
+export function alphabeticallySortedProductOptions(
+	families: APIProductFamily[]
+): APIProductFamilyProduct[] {
+	return sortBy( selectProductOptions( families ), ( product ) => product.name );
+}
```
Comment (heydemoura) [inline client/jetpack-cloud/sections/partner-portal/utils.ts:194]: Auxiliary functions for license issuing have been moved from the component level to this utils module.

File: client/state/partner-portal/licenses/hooks/use-issue-license-mutation.ts [hunk 12]
```diff
@@ -2,7 +2,7 @@ import { useMutation, UseMutationOptions, UseMutationResult } from 'react-query'
 import { wpcomJetpackLicensing as wpcomJpl } from 'calypso/lib/wp';
 import { APIError, APILicense } from 'calypso/state/partner-portal/types';
 
-interface MutationIssueLicenseVariables {
+export interface MutationIssueLicenseVariables {
 	product: string;
 }
```
Comment (heydemoura) [inline client/state/partner-portal/licenses/hooks/use-issue-license-mutation.ts:7]: Exporting interface for proper typing on the hook level

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+	product?: string | null
+): [
+	UseMutationResult< APILicense, APIError, MutationIssueLicenseVariables, unknown >,
+	boolean,
+	() => void
+] {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const sites = useSelector( getSites ).length;
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: Since we're modifying this, let's replace isSubmitting with a constant rather than state as its duplicating state that react-query already provides us with:
```
const isLoading = issueLicense.isLoading || assignLicense.isLoading.
```
using isLoading for consistency with react-query's naming.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+	const products = useProductsQuery( {
+		select: alphabeticallySortedProductOptions,
+	} );
+
+	const paymentMethodRequired = useSelector( doesPartnerRequireAPaymentMethod );
+	const hasPaymentMethod = useSelector( hasValidPaymentMethod );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: `doesPartnerRequireAPaymentMethod` already covers `hasPaymentMethod` so we can skip it.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+						page.redirect(
+							addQueryArgs(
+								{
+									return: addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) ),
+								},
+								partnerPortalBasePath( '/payment-methods/add' )
+							)
+						);
+						return;
+					}
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: If we no longer have the case where we try to issue a license without being sure if a payment method is available, we should probably remove this logical branch.

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 16]
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
...
-	// If a product is passed down from the query we want to instantly try and create a license for it
-	// and we only want to try that once hence why we pass no dependencies to useEffect().
-	// This is a short-term solution as we shouldn't be executing such actions with a GET request without a nonce.
-	useEffect( () => {
-		if ( defaultProduct !== '' ) {
-			page.redirect(
-				removeQueryArgs( window.location.pathname + window.location.search, 'product' )
-			);
-			issueLicense.mutate( { product: defaultProduct } );
+		if ( hasPaymentMethod ) {
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: We should use `! paymentMethodRequired` here instead otherwise this fails for non-agencies that don't have a payment method as they go the requirePaymentMethod() route which does nothing since a payment method is not required for non-agencies (i.e. non-agencies can't issue licenses right now :D).

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 17]
```diff
@@ -205,22 +68,15 @@ export default function IssueLicenseForm( {
 
 	const onIssueLicense = useCallback( () => {
 		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: We may want to move this even to the hook that handles issuing since we'd lose the event when the issuing happens when adding a payment method and issuing a license in one step.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+			// error is missing_valid_payment_method but our local state shows that the user has a payment method.
+			if ( ! paymentMethodRequired && error?.code === 'missing_valid_payment_method' ) {
+				return errorCount < 5;
+			}
+
+			return false;
+		},
+	} );
+
+	return [ issueLicense, isSubmitting, requirePaymentMethod ];
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: I think we should restructure this return value a bit like so:
```
const issue = useCallback( () => {
	if ( isLoading ) {
		return;
	}

	if ( paymentMethodRequired ) {
		const nextStep = addQueryArgs(
			{
				product,
			},
			partnerPortalBasePath( '/payment-methods/add' )
		);
		page( nextStep );
		return;
	}

	issueLicense.mutate( { product } );
}, [ product, paymentMethodRequired, issueLicense.mutate ] );

return [ issue, isLoading ];
```

This way callers of this hook don't have to worry about the payment method status and execute extra redirection functions - they only need to care about `isLoading` and to call `issue()`.
(isLoading comes from my earlier comment about replacing isSubmitting).

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx (hunk 6 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: useReturnUrl() already checks if the return query arg is empty so `!! returnQueryArg` is unnecessary I think.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx (hunk 7 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:116]: If I'm understanding this correctly, if there's only `product` but no `returnQueryArg` then we have nowhere to return to and refetching the cards to update `has_valid_payment_method` seems unnecessary since we won't redirect the user even if they have a new payment method i.e.  we should be fine with removing `|| product`.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 19]
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
 							<div className="payment-method-add__navigation-buttons">
 								<Button
 									className="payment-method-add__back-button"
-									href="/partner-portal/payment-methods/"
-									disabled={ isStripeLoading }
+									href={
+										product
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
+											: '/partner-portal/payment-methods/'
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Let's use `partnerPortalBasePath()` for new references to partner portal URLs i.e. `partnerPortalBasePath( '/payment-methods' )`.

Let's also move this conditional before rendering to make rendering easier to read.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 20]
```diff
@@ -186,8 +206,12 @@ function PaymentMethodAdd(): ReactElement {
 							<div className="payment-method-add__navigation-buttons">
 								<Button
 									className="payment-method-add__back-button"
-									href="/partner-portal/payment-methods/"
-									disabled={ isStripeLoading }
+									href={
+										product
+											? addQueryArgs( { product }, partnerPortalBasePath( '/issue-license' ) )
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:216]: Props for retaining some benefits from the last-minute implementation we had before so we could pre-select a product for the customer when they go back!

File: client/jetpack-cloud/sections/partner-portal/utils.ts [hunk 21]
```diff
@@ -176,3 +182,13 @@ export function formatApiPartner( partner: APIPartner ): Partner {
 export function getProductTitle( product: string ): string {
 	return product.replace( /(?:Jetpack\s|[)(])/gi, '' );
 }
+
+export function selectProductOptions( families: APIProductFamily[] ): APIProductFamilyProduct[] {
+	return families.flatMap( ( family ) => family.products );
+}
+
+export function alphabeticallySortedProductOptions(
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/utils.ts]: Let's rename this to make it obvious it's a query selector function like the one above:
`selectAlphabeticallySortedProductOptions`

**Review by atanas-dev — CHANGES_REQUESTED** [review]
Thank you for taking the lead on fixing our short-term cheating! 😆
Changes mostly look good - I've sprinkled comments around places where we can improve or simplify things.

Note that some of my comments may be on things that were just moved but since we're touching them now we might as well add some minor improvements to them.

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx (hunk 17 above)
Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: Smart! I've moved it to the hook level and replaced this event with `calypso_partner_portal_issue_selection_submit` for good measure, then we know if the user has moved on a step further.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx (hunk 18 above)
Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/hooks.tsx]: Yes, this is a far better approach. It reduces verbosity on every place we would need to issue a license.

File: client/jetpack-cloud/sections/partner-portal/controller.tsx [hunk 22]
```diff
@@ -242,10 +242,10 @@ export function requireSelectedPartnerKeyContext(
 export function requireValidPaymentMethod( context: PageJS.Context, next: () => void ) {
 	const state = context.store.getState();
 	const paymentMethodRequired = doesPartnerRequireAPaymentMethod( state );
-	const { pathname, search } = window.location;
+	const { path } = context;
 
 	if ( paymentMethodRequired ) {
-		const returnUrl = ensurePartnerPortalReturnUrl( pathname + search );
+		const returnUrl = ensurePartnerPortalReturnUrl( path );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/controller.tsx]: Is this change necessary? Ideally, this would redirect back to whatever the current URL was regardless of its query params.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+			}
+
+			return false;
+		},
+	} );
+
+	const isLoading = useMemo(
+		() => assignLicense.isLoading || issueLicense.isLoading,
+		[ issueLicense, assignLicense ]
+	);
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: This is overkill - `const isLoading = assignLicense.isLoading || issueLicense.isLoading;` is sufficient and faster since we're testing two booleans instead of involving function calls, memoization etc.

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+					product,
+				},
+				partnerPortalBasePath( '/payment-methods/add' )
+			);
+			page( nextStep );
+			return;
+		}
+
+		issueLicense.mutate( { product } );
+	}, [ product, paymentMethodRequired, issueLicense.mutate ] );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]:
```suggestion
	}, [ isLoading, product, paymentMethodRequired, issueLicense.mutate ] );
```

File: client/jetpack-cloud/sections/partner-portal/hooks.tsx [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+			);
+		},
+		onError: ( error: Error ) => {
+			dispatch( errorNotice( error.message ) );
+		},
+	} );
+
+	const issueLicense = useIssueLicenseMutation( {
+		onSuccess: ( license ) => {
+			dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/hooks.tsx]: Let's rename this to something like `calypso_partner_portal_license_issued ` since this is triggered after a license is successfully issued.

File: client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx [hunk 26]
```diff
@@ -204,23 +62,11 @@ export default function IssueLicenseForm( {
 		) );
 
 	const onIssueLicense = useCallback( () => {
-		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
-		issueLicense.mutate( { product } );
-	}, [ dispatch, product, issueLicense.mutate ] );
+		dispatch( recordTracksEvent( 'calypso_partner_portal_issue_selection_submit', { product } ) );
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/issue-license-form/index.tsx]: How about we move this to the hook that returns the issue function e.g.:
```
const issue = useCallback( () => {
	if ( isLoading ) {
		return;
	}
	dispatch( recordTracksEvent( 'calypso_partner_portal_issue_license_submit', { product } ) );
	// ...
```
This way the hook will handle this and not whatever is calling it every time.

Bonus: After this change we can remove the entire `onIssueLicense` handler and pass issueLicense to onClick directly.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 27]
```diff
@@ -127,7 +145,9 @@ function PaymentMethodAdd(): ReactElement {
 			<DocumentHead title={ translate( 'Payment Methods' ) } />
 			<SidebarNavigation />
 
-			{ returnQueryArg && <AssignLicenseStepProgress currentStep="addPaymentMethod" /> }
+			{ ( !! returnQueryArg || product ) && (
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:153]:
```suggestion
			{ ( !! returnQueryArg && !! product ) && (
```
I think we need both in order to consider this a part of the steps since no return = no flow / no next step and no product = no license issued.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx [hunk 28]
```diff
@@ -104,9 +116,15 @@ function PaymentMethodAdd(): ReactElement {
 				} )
 			);
 		} else {
-			page( '/partner-portal/payment-methods/' );
+			page( partnerPortalBasePath( '/payment-methods' ) );
+		}
+	}, [ page, product, issueLicense ] );
+
+	useEffect( () => {
+		if ( product && paymentMethodRequired ) {
```
Comment (atanas-dev) [inline client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: This doesn't work as expected - adding a payment method while issuing a license redirects to the payment methods listing page instead of to the assign license step.

After some debugging it seems this works:
```
const successCallback = useCallback( () => {
	if ( product ) {
		// Refresh stored cards to establish whether there is a valid payment method now.
		reduxDispatch(
			fetchStoredCards( {
				startingAfter: '',
				endingBefore: '',
			} )
		);
	} else {
		page( partnerPortalBasePath( '/payment-methods' ) );
	}
}, [ page, product, reduxDispatch ] );

useEffect( () => {
	if ( product && ! paymentMethodRequired ) {
		issueLicense();
	}
}, [ product, paymentMethodRequired ] );
```

The key changes here are:
- `if ( returnQueryArg ) {` -> `if ( product ) {`
- `if ( product && paymentMethodRequired ) {` -> `if ( product && ! paymentMethodRequired ) {`

Note that even with this change the process is a bit janky:
- The Go Back button is enabled during the process.
- We don't show a success notice for the license being issued - IMO we should make it clear that a license was issued since they are going to be charged for it even if they bail from the process before assigning the license.

We may want to address this in a separate task to make the experience smoother.

**Review by atanas-dev — CHANGES_REQUESTED** [review]
Things look a lot neater right now - props!

I've noted a few improvements but the major blocking piece is that the flow is broken right now as the user is shown the payment methods list instead of the assign license step after adding a payment method.

---

FYI I believe a12s will still be required to have a payment method because of a change introduced by Genesis where if the WPCOM user has an agency partner associated with it, even if that partner is not the current partner, the payment method requirement will apply because of how isAgencyUser() works.
If we decide to address this we can do so in a separate patch as it's not necessarily related to this one.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx (hunk 27 above)
Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:153]: Not exactly, they happen in different situations, not always TRUE together, and for those cases we would want to have the progress bar being rendered.

1. User has previously issued a license, deleted the payment methods, went to assign that license and was sent to payment method input. In this case there is no `product` parameter, just `return`, it would need to show the progress bar, as they are seemingly resuming the process.
2. User is issuing a license. There is no `return` parameter as they will be sent to a next step as a regular issuing process, but there's the `product` parameter.

Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx:153]: Makes sense 👍

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx (hunk 28 above)
Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Fixed the conditionals with the suggested changes, the redirect now works as expected.

Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: The process is a bit janky, we should make a separate task to make that smoother, which will involve less code and objective PRs for specific improvements.

Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Added a success notice for when the user successfully issues a license.

Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: I realized that we would need to test both `returnQueryArg` or `product`, as there's the case when a user has issued a license previously, had removed the payment method or it got expired, and will resume the process clicking on `Assign License` on the licenses list. This would send the user to the `payment-methods/add` page with a `return` query parameter, on that case we should not send the user to `/partner-portal/payment-methods` but send him to the license assignment (which is on the `return` query argument). Similar to what we have for rendering the progress bar.

So the approach would be:

```javascript
	const successCallback = useCallback( () => {
		if ( returnQueryArg || product ) {
			reduxDispatch(
				fetchStoredCards( {
					startingAfter: '',
					endingBefore: '',
				} )
			);
		} else {
			page( partnerPortalBasePath( '/payment-methods' ) );
		}
	}, [ page, product, reduxDispatch, partnerPortalBasePath, returnQueryArg ] );
```

File: client/jetpack-cloud/sections/partner-portal/controller.tsx (hunk 22 above)
Comment (heydemoura) [reply client/jetpack-cloud/sections/partner-portal/controller.tsx]: There was a bug that was fixed with other changes. So I'm reverting that change.

File: client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx (hunk 28 above)
Comment (atanas-dev) [reply client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx]: Nice catch 👍

**Review by atanas-dev — APPROVED** [review]
Patch tests good for me - great work! 🚢

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7495876

Thank you @heydemoura for including a screenshot in the description! This is really helpful for our translators.

**Comment by emilyaudela** [comment]
Hi there! It looks like there is a typo is this [string](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L201):

`Your %(product)s license was successfuly issued`

"sucessfuly" should be "successfully"

**Comment by heydemoura** [comment]
> Hi there! It looks like there is a typo is this [string](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L201):
> 
> `Your %(product)s license was successfuly issued`
> 
> "sucessfuly" should be "successfully"

Thanks for spotting this,
This typo is being handled on a different PR.
https://github.com/Automattic/wp-calypso/pull/67044

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7495876).
